### PR TITLE
[CLA] Add Damage CLs with Implex integration scheme

### PR DESCRIPTION
**Pull request by marandra**
**Description**
Add Implex version of symmetric and traction-only damage constitutive laws.

**Changelog**
- Added Implex version of small strain isotropic damage 3d constitutive law to ConstitutiveLawsApplication
- Added Implex version of traction-only small strain isotropic damage 3d constitutive law to ConstitutiveLawsApplication
- Exported to python
- Tests

### CODE REVIEW

File: applications/ConstitutiveLawsApplication/tests/cpp_tests/test_small_strain_isotropic_damage_implex_3d.cpp [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+    //  CREATE LAW
+    //
+
+    ConstitutiveLaw::Parameters cl_parameters;
+    Properties material_properties;
+    Vector stress_vector(6), strain_vector(6);
+    Matrix const_matrix;
+    // Create gauss point
+    Model current_model;
+    ModelPart& test_model_part = current_model.CreateModelPart("Main");
```
Comment (loumalouomega) [inline applications/ConstitutiveLawsApplication/tests/cpp_tests/test_small_strain_isotropic_damage_implex_3d.cpp:51]: There are utilities to generate this simple geometries, see TestUtilities

File: applications/ConstitutiveLawsApplication/tests/cpp_tests/test_small_strain_isotropic_damage_traction_only_implex_3d.cpp [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+    //  CREATE LAW
+    //
+
+    ConstitutiveLaw::Parameters cl_parameters;
+    Properties material_properties;
+    Vector stress_vector(6), strain_vector(6);
+    Matrix const_matrix;
+    // Create gauss point
+    Model current_model;
+    ModelPart& test_model_part = current_model.CreateModelPart("Main");
```
Comment (loumalouomega) [inline applications/ConstitutiveLawsApplication/tests/cpp_tests/test_small_strain_isotropic_damage_traction_only_implex_3d.cpp:51]: Idem

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Minor comments (i cannot review the theory)

**Comment by AlejandroCornejo** [comment]
Besides the comments from @loumalouomega is okkey for me. What it is this "implex" thing? :D

**Comment by loumalouomega** [comment]
Yep

**Comment by marandra** [comment]
> What it is this "implex" thing? :D

It's an implicit/explicit integration scheme (Implex) with adaptive time stepping algorithm.
From  Oliver, J., A.Huespe, and J.Cante. 2008. _"An implicit/explicit integration scheme to increase computability of non-linear material and contact/friction problems"_. Comput. Methods Appl. Mech. Eng.  197(21): 1865-1889:

"This work proposes a combination of implicit and explicit integration schemes that exploits the advantages of both, while overcoming some of their drawbacks. In essence, it is a combination of a standard implicit integration scheme of the stresses in the constitutive model in Eq. (3) with an explicit extrapolation of the involved internal variables in Eqs. (2)–(3)"

- "...The algorithmic tangent constitutive tensor becomes symmetric and semi-positive deﬁnite even in those cases as the analytical one is not. This leads to dramatic improvements of the robustness in problems where implicit integrations result in singularity or the negative character of the algorithmic tangent operators.
- In many cases, the algorithmic tangent constitutive tensor becomes constant. Therefore, in absence of sources of non-linearity other than the constitutive model, the complete non-linear problem reduces to a sequence of linear (at every time step) problems. The classical Newton–Raphson procedure takes a unique iteration to converge and the problem becomes step-linear. The eﬀects on the computational costs are also dramatic.
- The good stability properties of the implicit integration algorithm are inherited by the proposed IMPL-EX integration algorithm.
- The order of accuracy of the IMPL-EX integration algorithm, with respect to the size of the time step is, at least, linear; the same as many classical backward-Euler implicit algorithms. Nevertheless, the absolute error is larger for the same time step length.
- The method can be exploited to render robust, and step- linear, complex non-linear problems. For instance: in bi-lateral contact/friction problems, a strategy based on using an anticipating contact interface mesh, whose elements are equipped with an appropriate constitutive model, to penalize interpenetration and deﬁne the friction eﬀects, can be linearized by using an IMPL-EX integration of that constitutive model, resulting in a sequence of linear problems..."

**Comment by marandra** [comment]
> There are utilities to generate this simple geometries, see TestUtilities

@loumalouomega good to know. I am having problems finding the utilities in the code though. Can you point me to a test that uses them? It'll be easier to get it from a real example.

**Comment by loumalouomega** [comment]
> > There are utilities to generate this simple geometries, see TestUtilities
> 
> @loumalouomega good to know. I am having problems finding the utilities in the code though. Can you point me to a test that uses them? It'll be easier to get it from a real example. 

https://www.github.com/KratosMultiphysics/Kratos/tree/master/kratos%2Ftests%2Fcpp_tests%2Fprocesses%2Ftest_compute_nodal_gradient_process.cpp

Here is used for example, you can see the cpp utilities used and you can see the geometries generated in the source file

**Comment by loumalouomega** [comment]
> > > There are utilities to generate this simple geometries, see TestUtilities
> > 
> > @loumalouomega good to know. I am having problems finding the utilities in the code though. Can you point me to a test that uses them? It'll be easier to get it from a real example. 
> 
> https://www.github.com/KratosMultiphysics/Kratos/tree/master/kratos%2Ftests%2Fcpp_tests%2Fprocesses%2Ftest_compute_nodal_gradient_process.cpp
> 
> Here is used for example, you can see the cpp utilities used and you can see the geometries generated in the source file 

https://www.github.com/KratosMultiphysics/Kratos/tree/master/kratos%2Futilities%2Fcpp_tests_utilities.cpp

**Comment by marandra** [comment]
@loumalouomega I looked at the cpp test utilities, but I don't think it is relevant to these tests. In the test I don´t define elements, just one integration point for the CL. The 4 nodes are for defining a tetrahedron, as the geometry is required for some functions as Check and InitializeMaterial. If there are ways to improve these tests I'd be happy to discuss it, and as all the tests of the damage CLs are basically identical, it may make sense to modify all of them at once.

**Review by loumalouomega — APPROVED** [review]

**Comment by marandra** [comment]
Thanks!

**Pull request merged**